### PR TITLE
Add test for unexpected region for local data ReStatic

### LINKED ISSUE #49579: Unexpected region for local data ReStatic
```
#![feature(nll)]

fn fibs(n: u32) -> impl Iterator<Item=u128> {
    (0 .. n)
    .scan((0, 1), |st, _| {
        *st = (st.1, st.0 + st.1);
        Some(*st)
    })
    .map(&|(f, _)| f)
}

fn main() {
    println!("{:?}", fibs(10).collect::<Vec<_>>());
}
```

```
error: internal compiler error: unexpected region for local data ReStatic
 --> ...\test.rs:9:11
  |
9 |     .map(&|(f, _)| f)
  |           ^^^^^^^^^^

error: aborting due to previous error
```

```
rustc 1.26.0-nightly (06fa27d7c 2018-04-01)
binary: rustc
commit-hash: 06fa27d7c84a21af8449e06f3c50b243c4d5a7ad
commit-date: 2018-04-01
host: x86_64-pc-windows-gnu
release: 1.26.0-nightly
LLVM version: 6.0
```

### PR DESCRIPTION
**Pull request by davidtwco**
Fixes #49579.

r? @pnkfelix @nikomatsakis

### CODE REVIEW

File: src/test/ui/issue-49579.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+
+// compile-pass
+
+#![feature(nll)]
```
Comment (pnkfelix) [inline src/test/ui/issue-49579.rs]: Does this test *need* to utilize `#![feature(nll)]`? Or can we leave that of, and let the `compare-mode=nll` handle testing  that case (and thus get a test for this in both AST-borrowck and NLL modes)

Comment (davidtwco) [reply src/test/ui/issue-49579.rs]: This test does not pass without `#![feature(nll)]`. I'm not sure if it is possible to allow a UI test to compile successfully in one compare mode but fail in another.

Comment (pnkfelix) [reply src/test/ui/issue-49579.rs]: You can use `#[rustc_error]` to support that. There are examples of it elsewhere in the `ui/` test suite.

Comment (pnkfelix) [reply src/test/ui/issue-49579.rs]: but having said that, there's no reason to worry about it in this case. If the test does not compile w/o `#[feature(nll)]`, that is fine.

**Pull request merged**